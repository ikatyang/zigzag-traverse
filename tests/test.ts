import zigzag_traverse = require('../src/index');

test('throw error with non-integar height', () => {
  expect(() =>
    zigzag_traverse([0, 1, 2], 2, jest.fn()),
  ).toThrowErrorMatchingSnapshot();
});

test('does nothing with 0-length array', () => {
  const fn = jest.fn();
  zigzag_traverse([], 123, fn);
  expect(fn).not.toBeCalled();
});

test('5 x 4', () => {
  const results = visit(5, 4);
  expect(results.length).toBe(5 * 4);
  expect(results).toMatchSnapshot();
});

test('5 x 5', () => {
  const results = visit(5, 5);
  expect(results.length).toBe(5 * 5);
  expect(results).toMatchSnapshot();
});

function visit(width: number, height: number) {
  const offsets = [];
  for (let i = 0; i < width * height; i++) {
    offsets.push(i);
  }

  interface Position {
    line: number;
    column: number;
    offset: number;
  }

  const positions: Position[] = [];

  zigzag_traverse(offsets, width, offset =>
    positions.push({
      offset,
      line: Math.floor(offset / width),
      column: offset % width,
    }),
  );

  return positions;
}
