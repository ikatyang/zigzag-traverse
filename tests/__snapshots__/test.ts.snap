// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`5 x 4 1`] = `
Array [
  Object {
    "column": 0,
    "line": 0,
    "offset": 0,
  },
  Object {
    "column": 1,
    "line": 0,
    "offset": 1,
  },
  Object {
    "column": 0,
    "line": 1,
    "offset": 5,
  },
  Object {
    "column": 0,
    "line": 2,
    "offset": 10,
  },
  Object {
    "column": 1,
    "line": 1,
    "offset": 6,
  },
  Object {
    "column": 2,
    "line": 0,
    "offset": 2,
  },
  Object {
    "column": 3,
    "line": 0,
    "offset": 3,
  },
  Object {
    "column": 2,
    "line": 1,
    "offset": 7,
  },
  Object {
    "column": 1,
    "line": 2,
    "offset": 11,
  },
  Object {
    "column": 0,
    "line": 3,
    "offset": 15,
  },
  Object {
    "column": 1,
    "line": 3,
    "offset": 16,
  },
  Object {
    "column": 2,
    "line": 2,
    "offset": 12,
  },
  Object {
    "column": 3,
    "line": 1,
    "offset": 8,
  },
  Object {
    "column": 4,
    "line": 0,
    "offset": 4,
  },
  Object {
    "column": 4,
    "line": 1,
    "offset": 9,
  },
  Object {
    "column": 3,
    "line": 2,
    "offset": 13,
  },
  Object {
    "column": 2,
    "line": 3,
    "offset": 17,
  },
  Object {
    "column": 3,
    "line": 3,
    "offset": 18,
  },
  Object {
    "column": 4,
    "line": 2,
    "offset": 14,
  },
  Object {
    "column": 4,
    "line": 3,
    "offset": 19,
  },
]
`;

exports[`5 x 5 1`] = `
Array [
  Object {
    "column": 0,
    "line": 0,
    "offset": 0,
  },
  Object {
    "column": 1,
    "line": 0,
    "offset": 1,
  },
  Object {
    "column": 0,
    "line": 1,
    "offset": 5,
  },
  Object {
    "column": 0,
    "line": 2,
    "offset": 10,
  },
  Object {
    "column": 1,
    "line": 1,
    "offset": 6,
  },
  Object {
    "column": 2,
    "line": 0,
    "offset": 2,
  },
  Object {
    "column": 3,
    "line": 0,
    "offset": 3,
  },
  Object {
    "column": 2,
    "line": 1,
    "offset": 7,
  },
  Object {
    "column": 1,
    "line": 2,
    "offset": 11,
  },
  Object {
    "column": 0,
    "line": 3,
    "offset": 15,
  },
  Object {
    "column": 0,
    "line": 4,
    "offset": 20,
  },
  Object {
    "column": 1,
    "line": 3,
    "offset": 16,
  },
  Object {
    "column": 2,
    "line": 2,
    "offset": 12,
  },
  Object {
    "column": 3,
    "line": 1,
    "offset": 8,
  },
  Object {
    "column": 4,
    "line": 0,
    "offset": 4,
  },
  Object {
    "column": 4,
    "line": 1,
    "offset": 9,
  },
  Object {
    "column": 3,
    "line": 2,
    "offset": 13,
  },
  Object {
    "column": 2,
    "line": 3,
    "offset": 17,
  },
  Object {
    "column": 1,
    "line": 4,
    "offset": 21,
  },
  Object {
    "column": 2,
    "line": 4,
    "offset": 22,
  },
  Object {
    "column": 3,
    "line": 3,
    "offset": 18,
  },
  Object {
    "column": 4,
    "line": 2,
    "offset": 14,
  },
  Object {
    "column": 4,
    "line": 3,
    "offset": 19,
  },
  Object {
    "column": 3,
    "line": 4,
    "offset": 23,
  },
  Object {
    "column": 4,
    "line": 4,
    "offset": 24,
  },
]
`;

exports[`throw error with non-integar height 1`] = `"Expected height to be an integer, but received 1.5"`;
